### ISSUE TITLE
External field

### ISSUE THREAD

**Issue by Tissot11** [SmileiPIC/Smilei#283]
## Description

I'm trying to initialise external magnetic field using trapezoidal profile. However, at the left boundary of the simulation, I see a ramp in magnetic field strength despite I have set xslope1 and xslope2 to be zero. On the right boundary I don't see any ramping but only on the right side. I attach a snapshot and 
![Figure_8](https://user-images.githubusercontent.com/31315017/89496890-7df0a300-d7bb-11ea-8ce0-3e43209ea826.png)
paste some lines of namelist below. I'm also injecting particles but the issue using the random initialisation as suggested #271. 

```
EM_boundary_conditions = [ ['silver-muller','silver-muller'] ] ,

ExternalField(
    field = "Bx",
    profile = trapezoidal(0.03548, xvacuum=0.0,xplateau=400.*l0,xslope1=0.0,xslope2=0.0)
)
ExternalField(
    field = "Bz",
    profile = trapezoidal(0.03548, xvacuum=0.0,xplateau=400.*l0,xslope1=0.0,xslope2=0.0)
)
ExternalField(
    field = "Bx",
    profile = trapezoidal(0.0357,xvacuum=400.*l0, xplateau=Lsim,xslope1=0.0,xslope2=0.0,)
)
ExternalField(
    field = "Bz",
    profile = trapezoidal(0.1401,xvacuum=400.*l0, xplateau=Lsim,xslope1=0.0,xslope2=0.0,)
)
```

## Steps to reproduce the problem

If relevant, provide a step-by-step guide

## Parameters

- Smilei 4.3.2
- Input file ?
- Commands or script used to run the faulty simulation ?
- Computing resources
  - type of computer
  - OS
  - C++ compiler (g++ 4.8.5)
  - MPI version (g++ 6.2.0)
  - HDF5 version 1.8.21
  - Python version (`2.7.5`)
  - ...

**Comment by mccoys** [comment]
You see a ramp between the first 2 points of the grid. This is the limit of your resolution.

**Comment by Tissot11** [comment]
Thanks for the quick reply! Yes, I use a resolution of dx=0.1. But I would have thought that B should be constant at the boundary also. What is the best way to deal with this? Should a define xvacuum=-dx so that the field starts at a constant value at the left boundary? This ramping generates an additional component of the electric field causing unphysical results.

**Comment by mccoys** [comment]
If you want a constant field even past the boundary, you should set xvacuum to a large negative value, and let the plateau be much longer than the box. In the middle of the box, there is nothing you can do. It is impossible to have shorter changes then the resolution.

**Comment by Tissot11** [comment]
Ok. I could get away from the ramp on the left side by defining a python function profile. Is there any way to smoothen the fluctuation electric field that might arise due to this ramp in magnetic field in the middle of the simulation box?

**Comment by mccoys** [comment]
There are some filters available, but beware it might create unrealistic fields.

**Comment by Tissot11** [comment]
Ok. I'll check and compare the results. Friedman filter is only available in 2D? I follow the strategy of Plotnikov et al. (2017).

**Comment by Tissot11** [comment]
I have checked with filters and it made things even worse with unphysical results. I have another question related to the particle injector.  I initialized the injector based on the discussions in #271, and I attach the namelist. I have a plasma in the simulation box and I'm also injecting these electrons and ions from the left boundary for the continuous flow. The number density is fixed at 1.0, but I see it is not 1.0  (see the figures attached) at the left boundary and shows large fluctuations. Why is this the case? I have played around with different strategies and used trapezoidal fuction with xvacuum to be large negative number. But with the particle injector block on, it always shows unusual behaviour (Ex is also not zero). Am I not understanding how the particle injector block is meant to be used. Please let me know. 

I'm running the version of Smilei, I fetched from GitHub since the official tarball releases (4.4 and 4.5) I couldn't compile.


![Figure_1](https://user-images.githubusercontent.com/31315017/90195068-e3025500-ddc8-11ea-811a-7d95786499eb.png)
![Figure_2](https://user-images.githubusercontent.com/31315017/90195074-e564af00-ddc8-11ea-9284-5ae5572d1f71.png)
![Figure_3](https://user-images.githubusercontent.com/31315017/90195081-e72e7280-ddc8-11ea-8f9f-f4420b072e6e.png)
[namelist.txt](https://github.com/SmileiPIC/Smilei/files/5071868/namelist.txt)

__

**Comment by Tissot11** [comment]
What is also surprising that when I look at the phase space and I try to zoom in on the left boundary, I see rather unexpected behaviour. When I use the full grid, it looks reasonable.

![Figure_4](https://user-images.githubusercontent.com/31315017/90247236-7aea5800-de36-11ea-9c21-a2d03b557365.png)
![Figure_5](https://user-images.githubusercontent.com/31315017/90247285-8d649180-de36-11ea-85a7-c0a6b3409aa7.png)

**Comment by mccoys** [comment]
I am not very familiar with the injector, but I recall that particles are taken from a velocity distribution that may contain negative velocities. Those are discarded. This means there is a fluctuation due to the randomness of this sampling.

**Comment by Tissot11** [comment]
Thanks for the reply. Your explanation makes sense. The puzzling aspect was that I am initialising a supersonic flow, so i was expecting a lower number of particles with negative velocity. Anyway, is there any way to smooth the initial conditions in Smilei? I know about the current and field filtering but they work for the whole simulation duration. I could be interested in smoothening the initial conditions for the simulation. Do you think if this can be accomplished by using the restart feature of Smilei?

**Comment by Tissot11** [comment]
To suppress these fluctuations, I tried to do time averaging over 10 periods and saved the density in Field and the weight in ParticleBinning diagnostics. The results from two are different and I attach it here. How one can understand this?
![Field_Density](https://user-images.githubusercontent.com/31315017/90529464-0be16c00-e174-11ea-87dc-1c05c338a90d.png)
![ParticleBinning_Weight](https://user-images.githubusercontent.com/31315017/90529469-0c7a0280-e174-11ea-8a7a-7f79fe78b26b.png)

**Comment by mccoys** [comment]
I am not sure how you did your diagnostics here, and I am not sure what is the difference you want to highlight. In general, Field diagnostics obtain the density of particles using a projector like the one that is used to project the currents on the grid (usual PIC stuff). The particleBinning are very different, because they only compute a sum of particle weights in some user-defined binning.

In the future, please use a different issue for each question you have. In this issue you have asked many unrelated things and it is difficult to follow.

**Comment by Tissot11** [comment]
Sorry about these too many things. Some of these things are related and I'm not sure if one is not causing problems with others. I wanted to check is the profiles functions used to initialise plasma, fields and also use Particle Injector block. To summarize, I see few points that I don't understand.

1. I used your trick with trapezoidal functions that you had explained before for the external field and plasmas. It works well on the left boundary (for external fields but not for the density) but doesn't work on the right boundary for both fields and densities. I write below the lines I used and the figures to explain my point.

_ExternalField(
    field = "Bx",
    profile = constant(0.0358, xvacuum=-2*dx)
)
ExternalField(
    field = "Bz",
    profile = constant(0.140, xvacuum=-2*dx)
)_ 
For the _ExternalField_, I had also tried the trapezoidal function trick which I use below for the densities.

Number density for electrons and ions each

_number_density = trapezoidal(3.911, xvacuum=-2*dx,xplateau=40.0*Lsim,xslope1=0.0,xslope2=0.0),_

I also tried 

_number_density = trapezoidal(3.911, xvacuum=-2*dx,xplateau=40.0*Lsim,xslope1=0.0,xslope2=100.0),_

Please see the figures, below. In my case _Lsim=102.4_ and _dx=0.02_. On the right hand boundary, I see fields and density going to zero and I can not control it. For the density I used the _Probe_ diagnostic as it's much faster than _Fields_. I had used _Lsim_ for defining the corner for the _Probe_ diagnostic to calculate the densities in the figure.

![Bx](https://user-images.githubusercontent.com/31315017/91061188-2b710c80-e62c-11ea-9719-3fe352c0870c.png)
![Bz](https://user-images.githubusercontent.com/31315017/91061191-2c09a300-e62c-11ea-8eab-3705ab838463.png)
![Tot_dens](https://user-images.githubusercontent.com/31315017/91061192-2ca23980-e62c-11ea-9ffa-9152971b3e6d.png)

2. For density I thought it was happening due to the fluctuations cause by the _Particle Injector_ block, since even at the left boundary I can't have the correct initialised plasma density (see the figure)

![tot_dens_ini](https://user-images.githubusercontent.com/31315017/91065366-a89e8080-e630-11ea-833e-44a7961a8a4b.png)


So I tried saving density and fields with _Fields_  and _Probe_ diagnostics and also used _ParticleBinning_ for the density. I had used time averaging to avoid the fluctuation caused by the _Particle Injector_ block. This is what I wrote in the last post. The _ParticleBinning_ diagnostic gives correct value of particle density (1 in my case) but the _Field_ diagnostics does not give correct value of density (~0.10) if I use _time_average= 10_ for both diagnostics for whole simulation box.

To summarise, I want to know the reasons for the fields going to zero at time _t=0_ on the right boundary despite profile functions saying otherwise. And problem with the densities on both left and right hand boundaries. I also want to know if I save em fields using _Fields_ and _Probe_ diagnostics then I should get the same results or not.

**Comment by Tissot11** [comment]
Also does a constant external field is constant in time too?

**Comment by mccoys** [comment]
* Concerning profiles, have you tried a pure python function, like `lambda x, y: 0.140` ?

* I am not sure why the field diag gives 0.1. I will investigate.

* An external field is only applied at the beginning of the simulation. It does not have a time dependence.

**Comment by Tissot11** [comment]
1. I tried defining some basic Python functions based on conditional expressions. Again on the right side I had the same issue. I can try with lambda function tomorrow. I have also tried the polygonal function that you describe in the documentation.

2. Thanks for looking into it.

3.  What I see it that the external applied field shows strange behaviour at the left and right boundaries, which is unphysical. After running simulation for a while, on the left boundary it dies while it grows at the right boundary. In the remaining simulation box, it seems to show the physically meaningful behaviour. It should remain constant like at t=0 at left and right boundaries.

**Comment by mccoys** [comment]
1. The issue with profiles is actually not related to profiles ! It is due to the non-periodic box, and to the projection scheme. Let me explain. The projection of particle's quantities on the grid assumes some "spread" of each particle (the shape function), related to the order of interpolation (this is classic PIC stuff). When the plasma is uniform, the spreading of each particle will be exactly compensated by the spread of all surrounding particles, leading to uniform grid quantities. But when the plasma is not uniform (say, for instance, a jump in density) there will be a smoothing due to this spread. Now, when the box is not periodic, Smilei must assume there are no particles on the other side of the boundary, meaning it is a jump in density. The interpolation scheme will see it as a ramp in density. Hopefully, this density is only used in diagnostics, it has no effect on the physics. Now, I am not sure how this will appear for the currents as they do have an impact on the physics. I have seen that you have drifting particles in your box. @xxirii @MickaelGrechX do you know how to handle this boundary issue ?

2. Concerning the time_average issue, I just found out that this feature was not done for densities ! Sorry for the inconvenience. I will fix it soon.

3. Concerning external fields I cannot reproduce your problem. Could you provide an example file ?

**Comment by xxirii** [comment]
Dear @Tissot11, I have run your namelist to test the injection. I don't have the problem at the boundary. Is it something that happens at a given iteration ?

For instance, I have plotted the density from the particle binning diagnostics and the charge from the field diagnostics at timestep 1600:

![part_bin_1600](https://user-images.githubusercontent.com/17836646/91169763-29fd1e00-e6d8-11ea-8d3e-c2e9022d101d.png)
![rho_1600](https://user-images.githubusercontent.com/17836646/91169773-2e293b80-e6d8-11ea-8f48-f7fcad0b04d8.png)

Note that I agree with @mccoys explanation concerning fluctuations. Increasing the number of particle per cell at injection may improve the situation.

**Comment by Tissot11** [comment]
1. I understand that drifting plasmas can be tricky. If I remember correctly that I also tried last week periodic boundary conditions but some issues were still there. Anyway for particle density, the drift can be an issue as you explain but not for the external field applied in the simulation. They should not decay at the right boundary.

2. Thanks for the clarification.

3. Last night I ran a simplified set-up of the simulation I was running before. I attach the namelist here and some snapshots. To summarise I see issue with external field (_Bz_) at the right boundary at _t=0_ and later on both boundaries. It decays at the left and grows closer to the right boundary. I use _Probe_ and _ParticleBinning_ to sum up the densities of ions and electrons, I see difference in magnitude for in both diagnostics. But strangely this time at the left boundary and the reason of the difference is the ion density I see a big unphysical spike in ion density after _t=1000_.

First the namelist

[namelist.txt](https://github.com/SmileiPIC/Smilei/files/5123500/namelist.txt)

Bz profiles

![Bz_xmin](https://user-images.githubusercontent.com/31315017/91170337-07b7d000-e6d9-11ea-9dfa-2b125959b7e7.png)
![Bz_xmax](https://user-images.githubusercontent.com/31315017/91170348-0b4b5700-e6d9-11ea-9e68-9436b145860a.png)
![Bz_950](https://user-images.githubusercontent.com/31315017/91170360-0edede00-e6d9-11ea-967d-0a4cd4d62fa4.png)

Densities. First Probe and then ParticleBinning

![Rh_tot_Probe](https://user-images.githubusercontent.com/31315017/91170434-28802580-e6d9-11ea-85f4-a0a7846a7e33.png)
![Rho_tot_PB](https://user-images.githubusercontent.com/31315017/91170440-2ae27f80-e6d9-11ea-9e92-3dd8088e5005.png)

**Comment by Tissot11** [comment]
> Dear @Tissot11, I have run your namelist to test the injection. I don't have the problem at the boundary. Is it something that happens at a given iteration ?

This namelist I had used it to explain the problems with _time_average=10_ with both _ParticleBinning_ and _Fields_ diagnostics. But now this issue has been settled by @mccoys this morning. Please try my namelist that I have just sent and look at the results.

**Comment by xxirii** [comment]
In fact I have used the namelist you sent 12 days ago. I am only looking at the injection issue. I can see on my computer that the jump can be seen only in the rho diags as you mentioned not in the particle binning. It disappear around iteration 1000 but it may due to the particles coming back from the reflective wall. I think the cause is clearly the interpolation with the boundary behaviour. As soon as the physics is far from it I think it's ok.

**Comment by Tissot11** [comment]
Dear @xxirii, I have ran at least 30-40 simulations in different setups employing the particle injector block in last two weeks. Sorry that I mixed up _namelist_ before. I've not yet been able to analyze the simulations with a reflecting wall on the right. But please have a look at the _namelist_ that I have just attached 10 minutes before and also have a look at my results.

**Comment by mccoys** [comment]
Your issue is probably related to the particle boundary conditions, not external fields. The external fields are constant at the beginning so there is no problem. You see a drop on the left side only after a while. The drop at the right side is only caused by the fact that the point ``x=Lsim`` is considered outside of the box, so no field is evaluated there (it is always set to 0).

**Comment by Tissot11** [comment]
Sorry, I'm not able to fully follow you. At _t=0_ there is a problem on right boundary for the external field even though I have defined _xplateau=40*Lsim_ following your earlier suggestion. After _t>500_ one starts seeing decline of _Bz_ at the left while it grows at the right boundary, which can't be explained physically. I had also tried using _corners=Lsim-dx_ for the _Probe_ but it did help at _t=0_ but it manifested in a large spike in magnetic field at the right boundary and zero plasma density at later times. After a while density spikes up at the left boundary  which is clearly unphysical. I'm using _remove_ boundary conditions for the particle and _silver_muller_ for EM fields.

I'll try to run a simulation without plasmas and _Particle Injector_ block and apply an external field to see what happens.

**Comment by mccoys** [comment]
* at t=0 : I told you there is no problem at the right boundary. Your last probe point is outside of the box, so it does not record any field.

* at t > 500: You are seeing physical effects. I think that your B-field is transported by the plasma, but this is not caused by the code. You cannot initialize a B-field on particles that have not entered the box yet ! I am sure that this effect is removed if your plasma is not drifting.

**Comment by Tissot11** [comment]
> Dear @Tissot11, I have run your namelist to test the injection. I don't have the problem at the boundary. Is it something that happens at a given iteration ?
> 
> For instance, I have plotted the density from the particle binning diagnostics and the charge from the field diagnostics at timestep 1600:
> 
> ![part_bin_1600](https://user-images.githubusercontent.com/17836646/91169763-29fd1e00-e6d8-11ea-8d3e-c2e9022d101d.png)
> ![rho_1600](https://user-images.githubusercontent.com/17836646/91169773-2e293b80-e6d8-11ea-8f48-f7fcad0b04d8.png)
> 
> Note that I agree with @mccoys explanation concerning fluctuations. Increasing the number of particle per cell at injection may improve the situation.

Now I looked it these results. Since I haven't put any external fields in this simulation, I only saw that at the left and right boundaries initialization problems. I attach my figures. One can see that at both boundaries the density is around half of what it should be.
![Rho_eon_xmin](https://user-images.githubusercontent.com/31315017/91178327-45baf100-e6e5-11ea-8edb-a06bd0aa3298.png)
![Rho_ion_xmin](https://user-images.githubusercontent.com/31315017/91178330-46538780-e6e5-11ea-93ba-1d23017e09b3.png)

**Comment by mccoys** [comment]
> One can see that at both boundaries the density is around half of what it should be.

I answered above this question. It is due to the projection algorithm that assumes there is no particle after the boundary. You have half the value because it makes some sort of smoothing between the plasma and the vacuum outside of the box.

**Comment by Tissot11** [comment]
> * at t=0 : I told you there is no problem at the right boundary. Your last probe point is outside of the box, so it does not record any field.

I have done few more simulations where I didn't use _Probe_ but only Fields to save density and fields and also choosing _corners=Lsim-dx_. I'll send you figures later after I sort these results out.

How about the spike at the left boundary for ions after _t=500_? Is this not unphysical?

> * at t > 500: You are seeing physical effects. I think that your B-field is transported by the plasma, but this is not caused by the code. You cannot initialize a B-field on particles that have not entered the box yet ! I am sure that this effect is removed if your plasma is not drifting.

Physically, it seems a bit improbable. Because the field is supposed to be initialised at the left boundary and it is constant throughout the simulation at least at the left boundary. In some cases, I can clearly see that it the same effect that you have particle injections. The value initialzed at the left boundary is a bit lower or even zero and somehow because of the drifting plasma, it starts moving to the right. This causes this field to decay. There is no transportation of magnetic field happening that fast on this time scale. I attach some figures to show what I meant. 

![B1_t0](https://user-images.githubusercontent.com/31315017/91185521-3a1ff800-e6ee-11ea-8b13-f22992c604b1.png)
![B1_t2](https://user-images.githubusercontent.com/31315017/91185526-3ab88e80-e6ee-11ea-8700-e01a17211071.png)
![B1_t3](https://user-images.githubusercontent.com/31315017/91185527-3b512500-e6ee-11ea-8560-3bd1d5d420ca.png)
![B1_t6](https://user-images.githubusercontent.com/31315017/91185528-3b512500-e6ee-11ea-9e43-059ba4a57c5b.png)
![B1_t13](https://user-images.githubusercontent.com/31315017/91185530-3be9bb80-e6ee-11ea-8d68-41892472914f.png)


In this particular simulation, when the _Bz_ becomes stronger at the right boundary, it stops the particle injection on the right boundary after 2e5 iterations. So this is contrary to what you're saying.  I attach the namelist and these results on the electron plasma density.

[namelist_9.txt](https://github.com/SmileiPIC/Smilei/files/5124260/namelist_9.txt)


![eon1](https://user-images.githubusercontent.com/31315017/91186601-8a4b8a00-e6ef-11ea-86ba-ac6b21047bd5.png)
![eon2](https://user-images.githubusercontent.com/31315017/91186605-8ae42080-e6ef-11ea-87b3-3c55ffee6185.png)


My question would be that if all these projection algorithms can also assume that there is a plasma in ghost cells behind the left boundary. The number of these ghost cells can be increased to get rid of these fluctuations at the boundaries to initialise both particles and external fields with desired values?

**Comment by Tissot11** [comment]
![Bz](https://user-images.githubusercontent.com/31315017/91188141-3fcb0d00-e6f1-11ea-98de-6d7172ca473d.png)

This is how the magnetic field looks

**Comment by mccoys** [comment]
> How about the spike at the left boundary for ions after t=500? Is this not unphysical?

I do not see a spike for ions

> Physically, it seems a bit improbable. Because the field is supposed to be initialised at the left boundary and it is constant throughout the simulation at least at the left boundary. 

I don't understand how the field should be constant throughout the simulation. It is only initialised at some value, then evolves according to Maxwell's equations.

> The value initialzed at the left boundary is a bit lower or even zero and somehow because of the drifting plasma, it starts moving to the right. This causes this field to decay. There is no transportation of magnetic field happening that fast on this time scale.

Why shouldn't there be transportation of magnetic field on that time-scale ? Doesn't it goes with the speed of the plasma ?

> My question would be that if all these projection algorithms can also assume that there is a plasma in ghost cells behind the left boundary.

Good question. I don't know if there has been some work done in that regard before.

> The number of these ghost cells can be increased to get rid of these fluctuations at the boundaries to initialise both particles and external fields with desired values? In other words what one should do to correctly initialise a constant external field and inject a constant flux of particles with a given density at the boundaries during the whole simulation time?

I don't think it has been tried !

**Comment by Tissot11** [comment]
> > How about the spike at the left boundary for ions after t=500? Is this not unphysical?
> 
> I do not see a spike for ions

I attach again the figure and a 2D line out of it.
![Rho_ion2-PB](https://user-images.githubusercontent.com/31315017/91199632-65aade80-e6fe-11ea-8c60-20678f161bee.png)
![ion2-PB](https://user-images.githubusercontent.com/31315017/91199635-66dc0b80-e6fe-11ea-9099-0af328129767.png)
 Please see at the left boundary. The ion density injected got too much suddenly.

> 
> > Physically, it seems a bit improbable. Because the field is supposed to be initialised at the left boundary and it is constant throughout the simulation at least at the left boundary.
> 
> I don't understand how the field should be constant throughout the simulation. It is only initialised at some value, then evolves according to Maxwell's equations.

If you look at those figures I sent before, you can clearly see that magnetic field drops to zero at the left boundary before anything has actually happened in the plasma in the simulation box.  Which I can not understand. Of course, I agree with you that in the simulation box where interaction happens between electrons and ions, this magnetic field will be affected after some time has elapsed.

.> 
> > The value initialzed at the left boundary is a bit lower or even zero and somehow because of the drifting plasma, it starts moving to the right. This causes this field to decay. There is no transportation of magnetic field happening that fast on this time scale.
> 
> Why shouldn't there be transportation of magnetic field on that time-scale ? Doesn't it goes with the speed of the plasma ?

The plasma we inject is neutral because I choose the drift velocities to cancel _u cross B_  which is the electric field or current in MHD. So it can not create a current to nullify the external magnetic field only at the left boundary. If you think in terms of magnetic field being transported with drifting plasma at the left boundary then at every instant of time, there should be a magnetic field present at the left boundary for the new particles being injected. This is essentially the shock simulation geometry. One injects the particles from one side and these particles are in a magnetic field always like in the upstream part of the shock. I hope now you understand my point.

> 
> > My question would be that if all these projection algorithms can also assume that there is a plasma in ghost cells behind the left boundary.
> 
> Good question. I don't know if there has been some work done in that regard before.
> 
> > The number of these ghost cells can be increased to get rid of these fluctuations at the boundaries to initialise both particles and external fields with desired values? In other words what one should do to correctly initialise a constant external field and inject a constant flux of particles with a given density at the boundaries during the whole simulation time?
> 
> I don't think it has been tried !

I have only used SMILEI and EPOCH, and now I've been using SMILEI most of the time for my work. Other codes such as TRISTAN and OSIRIS have had particle injectors for a while and I do not see any such behaviour in papers written using them. I had this intuitive idea based on ghost cells to suggest this. Of course I'm not an expert on PIC algorithms, so I leave it for your better judgement! I'm only interested  in getting this work very well in SMILEI without any issues. Thanks for looking into all these issues!

**Comment by Tissot11** [comment]
I just ran two simulations in vacuum where I only applied the external field. In Namelist I still keep the plasma species but put the _number_density=0_ while in the other I removed _Species_ block altogether. I attach both Namelists and the streak snapshots of _Bz_ and _Bx_. I'm surprised to see again problems with the _Bz_. Please have a look.

[vacuum.txt](https://github.com/SmileiPIC/Smilei/files/5125867/vacuum.txt)


![Bx](https://user-images.githubusercontent.com/31315017/91218185-fbebfe00-e718-11ea-8d11-d7aa833c5d8e.png)
![Bz](https://user-images.githubusercontent.com/31315017/91218193-fc849480-e718-11ea-9301-8aa7a9823c06.png)

[vacuum1.txt](https://github.com/SmileiPIC/Smilei/files/5125866/vacuum1.txt)


![Bx](https://user-images.githubusercontent.com/31315017/91218254-102ffb00-e719-11ea-9022-ef5c8e63d5c0.png)
![Bz](https://user-images.githubusercontent.com/31315017/91218258-10c89180-e719-11ea-96b2-b3afd73cfe8c.png)

**Comment by mccoys** [comment]
Actually, you must put at least 3*dx !! If you simply put a constant it also works.

**Comment by Tissot11** [comment]
Yeah, now it works! Is there some reason behind this particular _2*dx_ value? I had tried first very large value both for the density and external fields. I noticed some issues with the particle injector, so I choose just _2*dx_. Do you have any other suggestions that I should take into account about regarding whatever I have asked till now?

**Comment by mccoys** [comment]
The default interpolator width (number of ghost cells) is 2.

Concerning the suggestions you ask, I am not sure. Are you talking about the injector setup ? I am really not used to the injection of particles, so I think some others should answer when they return from vacation.

**Comment by Tissot11** [comment]
Ok. Yeah, I have still few questions about the particle injector but I'll wait until someone replies. Thanks a lot for the help! I did many too many runs recently and I have to now understand these better.

**Comment by Tissot11** [comment]
Sorry again coming back to this external field question. For contact magnetic field it does work. But I wanted to create a jump in magnetic field in the middle of the simulation box. I used both _trapezoidal_ and my own python functions to define this magnetic field _Bz_. I see that due to this jump there is a numerical instability that actually affects the magnetic field applied. Since I don't apply any jump in _Bx_, it can not show any effect there. I had hoped this numerical instability arising due to jump is after a while leaves the simulation box since I use _silver-muller_ boundary conditions. This is all in vacuum. Do  you have any suggestions on this?

[Vacuum-Trial.txt](https://github.com/SmileiPIC/Smilei/files/5135689/Vacuum-Trial.txt)
![Bz-Trap](https://user-images.githubusercontent.com/31315017/91427247-e6342100-e85d-11ea-8061-307eac6c93ba.png)

**Comment by mccoys** [comment]
A jump of Bz in the x direction makes a curl of the magnetic field. It should create Ey field over time, and then a traveling wave. Do you expect the fields to remain constant ?

**Comment by Tissot11** [comment]
Yes, your explanation is correct. I expect the _Bz_ field to be constant at and closer to the boundaries. At the location of jump in the simulation box and elsewhere, it will change.

**Comment by mccoys** [comment]
I am not sure what is the problem at the left. Your picture shows it constant.

**Comment by Tissot11** [comment]
The magnitude gets lower. Until _x<=40_ It should have been 0.035. Due to this instability _Ey_, _Bz_ becomes much lower even at the left boundary in vacuum at later time.

**Comment by Tissot11** [comment]
I guess the solution lies in defining a python function and not using built-in spatial other than _constant_ for the external field?

**Comment by mccoys** [comment]
* You forgot to put xvacuum = -3*dx

* This is not an instability. It is correct physics. You simply have an EM wave traveling at the speed of light.

**Comment by Tissot11** [comment]
I thought it was only needed for the _constant_ function. So this _3*dx_ initialisation is needed if I use any built-in SMILEI functions for _Bz_ field? For _Bx_ it doesn't matter and even a _2*dx_ initialisation works because  _Bx_ doesn't have any jump. Am I correct in interpreting this?

Would a _3*dx_ initialisation also be required if I define a Python function like below?

_def BzProfile(x):
        return 0.035+0.5*0.105*(1+numpy.tanh((x-40.0)))_

In the simulations I have done with plasma inlcuded in the simulation, I find the behaviour of the em pulse due to discontinuity in middle according to the expectation. However, in those simulations I see a strange behaviour at the left boundary which is not connected with this discontinuity in the middle. I was using the built-in _trapezoidal_ function to define _Bz_. So if I try _-3*dx_ initilzation then hopefully I might not see any problem at the left boundary?

**Comment by mccoys** [comment]
* 3*dx is only needed in functions which implement this `xvacuum` variable. It means the profile is 0 before that limit.
* For Bx, it works because it does not create a curl of B field.
* 3*dx is not needed in a python function. It does not make sense anyways because there is no such xvacuum variable.
* I tried your namelist, and the problem was caused by Bz=0 at the boundary. You should either use a python profile that does not have 0 at the boundary, or use xvacuum of at least 3*dx

**Comment by Tissot11** [comment]
Ok. Thanks for tips!

**Comment by mccoys** [comment]
I tried Vacuum_trial

**Comment by Tissot11** [comment]
Ok. I had done another simulation where I had set _xvacuum=-100_. This caused also unphysical results. So I should only set _xvacuum=-3*dx_ and not too large value. Sometimes, I also see issue with the left boundary even if I defined a Python function like

_
def BzProfile(x):
    if (x <= 40*l0):
        return 0.035
    else: 
        return 0.140  
_

However, it is inconsistent and seems to depend also on other parameters of the simulation box. Regarding the _corners_ of _Probe_ diagnostic, I had intuitively thought that _Lsim_ should be the corner. But this isn't case as you mentioned. Should I alway set _0.99*Lsim_ for it to work?

**Comment by mccoys** [comment]
* setting xvacuum to a large negative value, or using a python function works. If you had a problem, it was due to something else.

* The corners of a Probe can be set to whatever you like. If some points are outside the box, they will be set to 0. If no points are in the box, an error is raised. I don't know why you had an error, but you need to carefully choose your corners and the number of points so that they are in the box. You can have more probe points than grid points if you like, but this is not very helpful usually.

**Comment by Tissot11** [comment]
1. Yeah the simulation with _xvacuum=-100_ was the one I spoke few days ago and it could have something to do with the _ParticleInjector_ block. So I’ll wait until colleagues who wrote the _ParticleInjector_ block come back from holidays. 

2.  I had defined _Lsim-dx_ as the _corners_ and chose _Nx=Lsim/dx=102400_ to be points of the Probe.  Then it showed error about no points being in the domain. I  then reduced the points to _Nx=1000_ and the simulation is running fine.

**Comment by mccoys** [comment]
If you have an issue with probes, please open a new issue with more details on your setup.

**Comment by Tissot11** [comment]
I have opened the issue and attached the _namelist_ and _stdout_ file.

**Comment by mccoys** [comment]
Could you recall what was the question with the injector?  I am a bit lost

**Comment by Tissot11** [comment]
So I had two issues. 

1. I see that particle injectors seems to  inject less or few particles after a certain time on the right boundary but on the left boundary there is a sudden spike. Please have a look at these pictures

![91199635-66dc0b80-e6fe-11ea-9099-0af328129767](https://user-images.githubusercontent.com/31315017/92202290-f2803580-ee7e-11ea-95e9-903ee5883785.png)
![91199632-65aade80-e6fe-11ea-8c60-20678f161bee](https://user-images.githubusercontent.com/31315017/92202295-f449f900-ee7e-11ea-9e14-5589d7a42025.png)

2. I suspect that the fluctations in the injected particle density are sometimes causing an issue on the applied magnetic _Bz_ at the left boundary. Like here
![91185530-3be9bb80-e6ee-11ea-8d68-41892472914f](https://user-images.githubusercontent.com/31315017/92202433-4db22800-ee7f-11ea-9e2a-9ee441f1b660.png)

This was the namelist

[namelist_9.txt](https://github.com/SmileiPIC/Smilei/files/5172831/namelist_9.txt)


I'm running the same simulation again with higher resolution to see if it had any impact on the results. Thanks for looking in this issue!

**Comment by mccoys** [comment]
* Be careful with profiles at boundaries, as explained before.

* You have positive velocity for species injected on the right. Is this really correct ?

* You have fields and several species. How can you expect them to go smoothly all the way ??? I don't have time to analyse the physics of your system, but you seem to have simply physical effect you need to understand.

**Comment by Tissot11** [comment]
Yes, I have positive drift for the plasma and I agree that it's a complicated system but I am trying to reproduce results that have been published before using other codes. This is the reason that I suggested that there can be some problems. I have done some test with stationary and drifting plasmas (without particle injections) and external fields and things look reasonable.

**Comment by xxirii** [comment]
Hi @Tissot11,

I have tested the last version of your namelist on my small computer until time t = 300. There is indeed something strange. I can observe as well a spike growing at the right boundary. Particle injection may not be adapted in your case using an external field near the boundaries. I am not sure injecting particles as we do is correct in this case. Behind the boundaries, the magnetic field do not exist and the plasma is thermalized. In principle, the physics should happen far from the boundaries.

Do you have examples in the literature of people doing the same thing with an injector ?

**Comment by Tissot11** [comment]
Thanks for the quick answer! I'm also a bit puzzled and I would like to understand the source of this problem. I attach some papers here. My setup is similar but now I'm trying to reproduce these exact results. I attach two papers here. Please have a look.

[Full particle simulation of a perpendicular collisionless shock- A shock-rest-frame model-EarthPlanetsSpace_58_e41.pdf](https://github.com/SmileiPIC/Smilei/files/5175273/Full.particle.simulation.of.a.perpendicular.collisionless.shock-.A.shock-rest-frame.model-EarthPlanetsSpace_58_e41.pdf)

[Dynamics-and-microinstabilities-at-perpendicular-collisionless-shock-A-comparison-of-largescale-twodimensional-full-particle-simulations-with-different-ion-to-electron-mass-ratio2014Physics-of-Plasmas.pdf](https://github.com/SmileiPIC/Smilei/files/5175286/Dynamics-and-microinstabilities-at-perpendicular-collisionless-shock-A-comparison-of-largescale-twodimensional-full-particle-simulations-with-different-ion-to-electron-mass-ratio2014Physics-of-Plasmas.pdf)

**Comment by xxirii** [comment]
I see, I have to look at the injection methods maybe similar to what was suggested in #271. This is a long work and I cannot focus on this for the moment. I will try at least to get more information about the methods when I can.

**Comment by Tissot11** [comment]
I had tried your workaround that you suggested in #271. I understand that you may have other things on your desk. It would be nice to have an understanding of the issue. Why there is an accumulation of particles on the right and sometime on the left boundaries, and why it seems to affect the external field on the left boundary.

**Comment by mccoys** [comment]
I am closing this issue as #293 is more specific on the remaining problems.
